### PR TITLE
Store Questions Bank progress in db

### PR DESCRIPTION
**Pull request by bognix**

### CODE REVIEW

File: app/Http/Controllers/Api/PrivateApi/UserStateApiController.php [hunk 1]
```diff
@@ -105,25 +105,43 @@ public function incrementTime(Request $request, $user)
...
-	public function saveQuizPosition(Request $request, $user)
+	public function saveQuizPosition(Request $request)
 	{
-		$cacheKey = $this->hashedFilters($request->filters);
-		$cacheTags = $this->getFiltersCacheTags(config('papi.resources.quiz-questions'), $user);
-		Cache::tags($cacheTags)->put($cacheKey, $request->position, 60 * 24);
+		$key = $this->hashedFilters($request->filters);
+		$value = $request->position;
+		$state = UserQuestionsBankState::firstOrNew(
+			['user_id' => $request->route('id')]
```
Comment (gitkuba) [inline app/Http/Controllers/Api/PrivateApi/UserStateApiController.php]: How about some kind of validation? Just to make sure we won't try to save `0` or `undefined` as id?

File: app/Http/Controllers/Api/PrivateApi/UserStateApiController.php [hunk 2]
```diff
@@ -105,25 +105,43 @@ public function incrementTime(Request $request, $user)
...
+		if (!Auth::user()->can('update', $state)) {
+			return $this->respondForbidden();
+		}
+
+
+		$state->key = $key;
+		$state->value = json_encode($value);
+		$state->save();
 
 		return $this->json([
```
Comment (gitkuba) [inline app/Http/Controllers/Api/PrivateApi/UserStateApiController.php]: Can we use `respondOk` for consistency?

File: app/Http/Controllers/Api/PrivateApi/UserStateApiController.php [hunk 3]
```diff
@@ -105,25 +105,43 @@ public function incrementTime(Request $request, $user)
...
+			['user_id' => $request->route('id')]
+		);
+
+		if (!Auth::user()->can('update', $state)) {
+			return $this->respondForbidden();
+		}
+
+
+		$state->key = $key;
+		$state->value = json_encode($value);
```
Comment (gitkuba) [inline app/Http/Controllers/Api/PrivateApi/UserStateApiController.php]: How about using json <-> array casting in model?

Comment (bognix) [reply app/Http/Controllers/Api/PrivateApi/UserStateApiController.php]: cast only works one way - on read, not both ways.

File: app/Http/Controllers/Api/PrivateApi/UserStateApiController.php [hunk 4]
```diff
@@ -105,25 +105,42 @@ public function incrementTime(Request $request, $user)
...
 	{
-		$cacheKey = $this->hashedFilters($request->filters);
-		$cacheTags = $this->getFiltersCacheTags(config('papi.resources.quiz-questions'), $user);
-		$cachedPosition = Cache::tags($cacheTags)->get($cacheKey, $request->position);
+		$key = $this->hashedFilters($request->filters);
+
+		$state = UserQuestionsBankState::select(['value', 'user_id'])
+			->where('user_id', $request->route('id'))
+			->where('key', $key)
+			->first();
```
Comment (rogatty) [inline app/Http/Controllers/Api/PrivateApi/UserStateApiController.php]: Shouldn't this be a model method?

File: app/Http/Controllers/Api/PrivateApi/UserStateApiController.php [hunk 5]
```diff
@@ -105,25 +105,42 @@ public function incrementTime(Request $request, $user)
...
+
+		if (!empty($state) && !Auth::user()->can('view', $state)) {
```
Comment (rogatty) [inline app/Http/Controllers/Api/PrivateApi/UserStateApiController.php]: What is `!empty($state)` for? Do we want to allow other users to know that someone has an empty state?

File: app/Http/Controllers/Api/PrivateApi/UserStateApiController.php [hunk 6]
```diff
@@ -105,25 +105,42 @@ public function incrementTime(Request $request, $user)
...
+			return $this->respondForbidden();
+		}
 
 		return $this->json([
-			'position' => $cachedPosition,
+			'position' => json_decode($state->value ?? '')
```
Comment (rogatty) [inline app/Http/Controllers/Api/PrivateApi/UserStateApiController.php]: Shouldn't this be an empty array or object?

**Pull request merged**